osbuild/buildroot.py

```diff
@@ -14,115 +19,169 @@
+            #
+            #   * We create a temporary directory for variable data and then use
+            #     it as '/var' in the container. This allows the container to
+            #     create throw-away data that it does not want to put into a
+            #     tmpfs.
+
+            api = tempfile.TemporaryDirectory(prefix="osbuild-api-", dir=self._rundir)
+            self.api = self._exitstack.enter_context(api)
+
+            dev = tempfile.TemporaryDirectory(prefix="osbuild-dev-", dir=self._rundir)
+            self.dev = self._exitstack.enter_context(dev)
+
+            var = tempfile.TemporaryDirectory(prefix="osbuild-var-", dir=self._vardir)
+            self.var = self._exitstack.enter_context(var)
+
+            subprocess.run(["mount", "-t", "tmpfs", "-o", "nosuid", "none", self.dev], check=True)
+            self._exitstack.callback(lambda: subprocess.run(["umount", "--lazy", self.dev], check=True))
+
+            self._mknod(self.dev, "full", 0o666, 1, 7)
+            self._mknod(self.dev, "null", 0o666, 1, 3)
+            self._mknod(self.dev, "random", 0o666, 1, 8)
+            self._mknod(self.dev, "urandom", 0o666, 1, 9)
+            self._mknod(self.dev, "tty", 0o666, 5, 0)
+            self._mknod(self.dev, "zero", 0o666, 1, 5)
+
+            self._exitstack = self._exitstack.pop_all()
+
+        return self
+
+    def __exit__(self, exc_type, exc_value, exc_tb):
+        with self._exitstack:
+            pass # Cleanup of the `ExitStack`.
 
     def run(self, argv, binds=None, readonly_binds=None):
         """Runs a command in the buildroot.
 
-        Its arguments mean the same as those for subprocess.run().
+        Takes the command and arguments, as well as bind mounts to mirror
+        in the build-root for this command.
         """
 
-        nspawn_ro_binds = []
-
-        # make osbuild API-calls accessible to the container
-        nspawn_ro_binds.append(f"{self.api}:/run/osbuild/api")
-
-        # We want to execute our stages and other scripts in the container. So
-        # far, we do not install osbuild as a package in the container, but
-        # provide it from the outside. Therefore, we need to provide `libdir`
-        # via bind-mount. Furthermore, a system-installed `libdir` has the
-        # python packages separate in `site-packages`, so we need to bind-mount
-        # them as well.
-        # In the future, we want to work towards mandating an osbuild package to
-        # be installed in the container, so the build is self-contained and does
-        # not take scripts from the host. However, this requires osbuild
-        # packaged for those containers. Furthermore, we want to keep supporting
-        # the current import-model for testing and development.
-        if self.libdir is not None:
-            # caller-specified `libdir` must be self-contained
-            nspawn_ro_binds.append(f"{self.libdir}:/run/osbuild/lib")
+        mounts = []
```

Comment:
How about a `if not self._exitstack: throw ValueError("Context not active")` or something?